### ISSUE TITLE
Netty - custom Channel initialization 

### ISSUE THREAD

**Issue by EliyahuSternDAI** [grpc/grpc-java#11149]
TL;DR: I want to send and receive custom data on the Netty Channel before the ManagedChannel functionality kicks-in and handles the HTTP/2 / gRPC protocols themselves. The aim is to first establish an ADB transport over the TPC connection to a socket on an Android device.

Am I developing gRPC in the following setup:
- The server is running inside an Android service, and is bound to a unix domain socket (I've implemented a `UdsServerSocket` similar to android/src/main/java/io/grpc/android/UdsSocket.java).
- The java client is running on a debugging host machine (where the device is connected to USB and can be connected over adb).

On the client (device host), I've been using `adb forward` to forward a TCP port to the device socket, which works fine.   

Since the `adb forward` setup introduces some problems, I want to connect from my java client directly to the ADB server, and send a few ADB messages (namely `host:transport:<device_id>` and `localabstract:<socket_path>`) and wait for a bunch of replies before I hand the connection over to the gRPC implementation.

I actually already have a library that handles creating such ADB transports on top of Netty Channels, and I was hoping to be able to use an already initialized Channel from my library for the gRPC client.

However,  searching the APIs for a few hours, I couldn't find a way to use my own Channels, or to add custom handlers into the pipeline.

As a side note, I am also working on a similar solution for iOS, and I expect to find myself in the same exact situation soon...

**Comment by ejona86** [comment]
> I want to send and receive custom data on the Netty Channel

Yeah... I don't think we want that in general. Things are brittle enough there, you'd highly likely be broken at some point. In general, we've added such things to this repo instead. You can already provide your own Channel factory. Having a custom channel is less of a problem than having additional handlers in the pipeline. In our repo, we make custom grpc-netty handshakers, but they are wholly unsupported outside of our repo.

Instead, we'd recommend configuring grpc-netty to use a LocalChannel, and then you can have a LocalServerChannel do any translation you need. Or using an okhttp channel with an adapter Socket implementation (as done with UdsSocketFactory in grpc-android).

I've not heard of problems with `adb forward`; I totally believe they exist, but I've not dealt with it much. I thought we had other users using it along with the OkHttpServer and the `grpcproxy` (example) for debugging/testing certain BinderServer services.

**Comment by EliyahuSternDAI** [comment]
> we'd recommend configuring grpc-netty to use a LocalChannel
That sounds like a good idea, I'll give it a try. Do you think that using `EmbeddedChannel` in a similar way should also be possible?

**Comment by ejona86** [comment]
`EmbeddedChannel` wouldn't work as well. We do use it in our tests some, and even then it has been fickle. There were some fixes to it years ago, but I don't remember how close it ended up. You'd have trouble because you need to get the instance of the channel, but that'd be hidden inside gRPC. LocalChannel handles that problem by having the global name-based registration.

**Comment by EliyahuSternDAI** [comment]
Ok, that makes sense.

BTW one problem with `adb forward` I wanted to avoid is that if the device get's disconnected (e.g. during reboot), gRPC will not be able to retry the connection/rpc until I set up the forward again on the same port (that may have, rarely, already been occupied by another process).
Also, forwards are globally controllable, therefore other processes (e.g. Android Studio) can tear them down for their own reasons.

I am hoping to find a solution that does not require my code to manually retry when the connection fail.

**Comment by ejona86** [comment]
Seems like this is resolved. If you encounter issues with the LocalChannel, comment, and it can be reopened.
